src/jarabe/desktop/schoolserver.py

```diff
@@ -127,39 +170,79 @@ def register_laptop(url=_REGISTER_URL):
-        server = xmlrpclib.ServerProxy(url)
     try:
-        data = server.register(sn, nick, uuid_, profile.pubkey)
-    except (xmlrpclib.Error, TypeError, socket.error):
-        logging.exception('Registration: cannot connect to server')
-        raise RegisterError(_('Cannot connect to the server.'))
-    finally:
-        socket.setdefaulttimeout(None)
-
-    if data['success'] != 'OK':
-        logging.error('Registration: server could not complete request: %s',
-                      data['error'])
-        raise RegisterError(_('The server could not complete the request.'))
-
-    settings.set_string('jabber-server', data['jabberserver'])
+        response = urllib2.urlopen(db_url)
+        server_html = response.read()
+    except (urllib2.URLError, urllib2.HTTPError):
+        logging.exception('Registration: cannot connect to xs-authserver')
+
+    if server_html and profile.pubkey in server_html:
+        new_registration_required = False
+        registered_laptops = re.findall(r'{.+}', server_html)
+        for laptop in registered_laptops:
+            if profile.pubkey in laptop:
+                string_for_json = laptop.replace("&#39;", "\"").replace(
+                    " u\"", "\"")
+                data = json.loads(string_for_json)
+                history_found, data_ = _get_history_for_serial(data["serial"])
+                new_registration_required = not history_found
+                if history_found:
+                    uuid_ = data_["uuid"]
+                    sn = data["serial"]
+                    backup_url = data_["backup_url"]
+                    try:
+                        jabber_server = re.search(r'\@(.*)\:',
+                                                  backup_url).group(1)
+                    except AttributeError:
+                        pass
+
+    if new_registration_required:
+        if sys.hexversion < 0x2070000:
+            server = xmlrpclib.ServerProxy(url, _TimeoutTransport())
+        else:
+            socket.setdefaulttimeout(_REGISTER_TIMEOUT)
+            server = xmlrpclib.ServerProxy(url)
+        try:
+            data = server.register(sn, nick, uuid_, profile.pubkey)
+        except (xmlrpclib.Error, TypeError, socket.error):
+            logging.exception('Registration: cannot connect to server')
+            raise RegisterError(_('Cannot connect to the server.'))
+        finally:
+            socket.setdefaulttimeout(None)
+
+        if data['success'] != 'OK':
+            logging.error('Registration: server could not complete request: %s',
+                          data['error'])
+            raise RegisterError(_('The server could not complete the request.'))
+
+        # Registration Successful, hence we can add the identification of this
+        # server to our known_host file.
+        command = "ssh-keyscan -H -t ecdsa " + jabber_server
+        output = subprocess.check_output(command, shell=True)
+        os.system("mkdir -p ~/.ssh")
```

Comment:
- .ssh directory may be missing. Hence I used -p.
- ya, i used ```os.system("mkdir -p ~/.ssh")``` instead of 
```if not os.path.exists(...):
    os.makedirs(...)```
for its simplicity. Shall i use the later?
- I didn't get the it. How would it break inbound SSH into the system?
- while calling ```ssh-keyscan``` we already know that the server is present and ```jabber_server``` value has correct address due to successful registration, I think assuming and checking for error will be redundant. The output of the command if executed successfully will either be the identity key or blank. that is why I didn't check it before, but now have added a condition to write only if output has value.
- I am not aware of the security of echoing to known_hosts. I replaced it ```with _write_to_file``` with append mode